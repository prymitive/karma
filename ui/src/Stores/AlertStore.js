import { observable, action, computed, toJS } from "mobx";

import throttle from "lodash.throttle";

import qs from "qs";

import { FetchGet } from "Common/Fetch";

const QueryStringEncodeOptions = {
  encodeValuesOnly: true, // don't encode q[]
  indices: false, // go-gin doesn't support parsing q[0]=foo&q[1]=bar
};

function FormatAlertsQ(filters) {
  return qs.stringify({ q: filters }, QueryStringEncodeOptions);
}

// generate URL for the UI with a set of filters
function FormatAPIFilterQuery(filters) {
  return qs.stringify(
    Object.assign(DecodeLocationSearch(window.location.search).params, {
      q: filters,
    }),
    QueryStringEncodeOptions
  );
}

// format URI for react UI -> Go backend requests
function FormatBackendURI(path) {
  const uri = process.env.REACT_APP_BACKEND_URI || ".";
  return `${uri}/${path}`;
}

// takes the '?foo=bar&foo=baz' part of http://example.com?foo=bar&foo=baz
// and decodes it into a dict with some extra metadata
function DecodeLocationSearch(searchString) {
  let defaultsUsed = true;
  let params = { q: [] };

  if (searchString !== "") {
    const parsed = qs.parse(searchString.split("?")[1]);
    params = Object.assign(params, parsed);

    if (parsed.q !== undefined) {
      defaultsUsed = false;
      if (parsed.q === "") {
        params.q = [];
      } else if (Array.isArray(parsed.q)) {
        // first filter out duplicates
        // then filter out empty strings, so 'q=' doesn't end up [""] but rather []
        params.q = parsed.q
          .filter((v, i) => parsed.q.indexOf(v) === i)
          .filter((v) => v !== "");
      } else {
        params.q = [parsed.q];
      }
    }
  }

  return { params: params, defaultsUsed: defaultsUsed };
}

function UpdateLocationSearch(newParams) {
  const baseURLWithoutSearch = window.location.href.split("?")[0];
  const newSearch = FormatAPIFilterQuery(newParams.q);
  window.history.pushState(
    null,
    null,
    `${baseURLWithoutSearch}?${newSearch || "q="}`
  );
}

const AlertStoreStatuses = Object.freeze({
  Idle: Symbol("idle"),
  Fetching: Symbol("fetching"),
  Processing: Symbol("processing"),
  Failure: Symbol("failure"),
});

function NewUnappliedFilter(raw) {
  return {
    applied: false,
    isValid: true,
    raw: raw,
    hits: 0,
    name: "",
    matcher: "",
    value: "",
  };
}

class AlertStore {
  filters = observable(
    {
      values: [],
      addFilter(raw) {
        if (this.values.filter((f) => f.raw === raw).length === 0) {
          this.values.push(NewUnappliedFilter(raw));
          UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
        }
      },
      removeFilter(raw) {
        if (this.values.filter((f) => f.raw === raw).length > 0) {
          this.values = this.values.filter((f) => f.raw !== raw);
          UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
        }
      },
      replaceFilter(oldRaw, newRaw) {
        const index = this.values.findIndex((e) => e.raw === oldRaw);
        if (index >= 0) {
          // first check if we would create a duplicated filter
          if (this.values.findIndex((e) => e.raw === newRaw) >= 0) {
            // we already have newRaw, simply drop oldRaw
            this.removeFilter(oldRaw);
          } else {
            // no dups, continue with a swap
            this.values[index] = NewUnappliedFilter(newRaw);
            UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
          }
        }
      },
      setFilters(raws) {
        this.values = raws.map((raw) => NewUnappliedFilter(raw));
        UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
      },
      setWithoutLocation(raws) {
        const filtersByRaw = this.values.reduce(function (map, obj) {
          map[toJS(obj.raw)] = toJS(obj);
          return map;
        }, {});
        this.values = raws.map((raw) =>
          filtersByRaw[raw] ? filtersByRaw[raw] : NewUnappliedFilter(raw)
        );
      },
      applyAllFilters() {
        for (let i = 0; i < this.values.length; i++) {
          this.values[i].applied = true;
        }
      },
    },
    {
      addFilter: action.bound,
      removeFilter: action.bound,
      replaceFilter: action.bound,
      setFilters: action.bound,
      setWithoutLocation: action.bound,
      applyAllFilters: action.bound,
    },
    { name: "API Filters" }
  );

  data = observable(
    {
      colors: {},
      counters: [],
      grids: [],
      silences: {},
      upstreams: { instances: [], clusters: {} },
      receivers: [],
      getAlertmanagerByName(name) {
        return this.upstreams.instances.find((am) => am.name === name);
      },
      isReadOnlyAlertmanager(name) {
        return this.readOnlyAlertmanagers.map((am) => am.name).includes(name);
      },
      getClusterAlertmanagersWithoutReadOnly(clusterID) {
        return this.clustersWithoutReadOnly[clusterID] || [];
      },
      get readOnlyAlertmanagers() {
        return this.upstreams.instances.filter((am) => am.readonly === true);
      },
      get readWriteAlertmanagers() {
        return this.upstreams.instances
          .filter((am) => am.readonly === false)
          .map((am) =>
            Object.assign({}, am, {
              clusterMembers: am.clusterMembers.filter(
                (m) => this.isReadOnlyAlertmanager(m) === false
              ),
            })
          );
      },
      get clustersWithoutReadOnly() {
        const clusters = {};
        for (const clusterID of Object.keys(this.upstreams.clusters)) {
          const members = this.upstreams.clusters[clusterID].filter(
            (member) => this.isReadOnlyAlertmanager(member) === false
          );
          if (members.length > 0) {
            clusters[clusterID] = members;
          }
        }
        return clusters;
      },
      getColorData(name, value) {
        if (this.colors[name] !== undefined) {
          return this.colors[name][value];
        }
      },
    },
    {
      readOnlyAlertmanagers: computed,
      readWriteAlertmanagers: computed,
      clustersWithoutReadOnly: computed,
    },
    { name: "API Response data" }
  );

  info = observable(
    {
      authentication: {
        enabled: false,
        username: "",
      },
      totalAlerts: 0,
      version: "unknown",
      upgradeNeeded: false,
      isRetrying: false,
      reloadNeeded: false,
      setIsRetrying() {
        this.isRetrying = true;
      },
      clearIsRetrying() {
        this.isRetrying = false;
      },
      setReloadNeeded() {
        this.reloadNeeded = true;
      },
    },
    {
      setIsRetrying: action.bound,
      clearIsRetrying: action.bound,
      setReloadNeeded: action,
    },
    { name: "API response info" }
  );

  settings = observable(
    {
      values: {
        staticColorLabels: [],
        annotationsDefaultHidden: false,
        annotationsHidden: [],
        annotationsVisible: [],
        sorting: {
          grid: {
            order: "startsAt",
            reverse: false,
            label: "alertname",
          },
          valueMapping: {},
        },
        silenceForm: {
          strip: {
            labels: [],
          },
        },
        alertAcknowledgement: {
          enabled: false,
          durationSeconds: 900,
          author: "karma / author missing",
          commentPrefix: "",
        },
      },
    },
    {},
    {
      name: "Global settings",
    }
  );

  status = observable(
    {
      value: AlertStoreStatuses.Idle,
      lastUpdateAt: 0,
      error: null,
      paused: false,
      setIdle() {
        this.value = AlertStoreStatuses.Idle;
        this.error = null;
        this.lastUpdateAt = new Date();
      },
      setFetching() {
        this.value = AlertStoreStatuses.Fetching;
      },
      setProcessing() {
        this.value = AlertStoreStatuses.Processing;
        this.error = null;
      },
      setFailure(err) {
        this.value = AlertStoreStatuses.Failure;
        this.error = err;
        this.lastUpdateAt = new Date();
      },
      pause() {
        this.paused = true;
      },
      resume() {
        this.paused = false;
      },
    },
    {
      setIdle: action,
      setFetching: action,
      setProcessing: action,
      setFailure: action,
      pause: action.bound,
      resume: action.bound,
    },
    { name: "Store status" }
  );

  constructor(initialFilters) {
    if (initialFilters !== null) this.filters.setFilters(initialFilters);
  }

  fetch = action(
    (gridLabel, gridSortReverse, sortOrder, sortLabel, sortReverse) => {
      this.status.setFetching();

      const args = [
        `gridLabel=${gridLabel}`,
        `gridSortReverse=${gridSortReverse ? "1" : "0"}`,
        `sortOrder=${sortOrder}`,
        `sortLabel=${sortLabel}`,
        `sortReverse=${sortReverse}`,
      ];

      const alertsURI =
        FormatBackendURI(`alerts.json?&${args.join("&")}&`) +
        FormatAPIFilterQuery(this.filters.values.map((f) => f.raw));

      return FetchGet(alertsURI, {}, this.info.setIsRetrying)
        .then((result) => {
          // we're sending requests with mode=cors so the response should also be type=cors
          // after a few failures in the retry loop we will switch to no-cors
          // if that request comes back as type=opaque then we might be getting
          // redirected by an auth proxy
          if (result.type === "opaque") {
            this.info.setReloadNeeded();
          }
          this.info.clearIsRetrying();
          this.status.setProcessing();
          return result.json();
        })
        .then((result) => {
          return this.parseAPIResponse(result);
        })
        .catch((err) => {
          console.trace(err);
          return this.handleFetchError(
            `Can't connect to the API, last error was "${err.message}"`
          );
        });
    }
  );

  fetchWithThrottle = throttle(this.fetch, 300);

  parseAPIResponse = action((result) => {
    if (result.error) {
      this.handleFetchError(result.error);
      return;
    }

    const queryFilters = [
      ...new Set(
        this.filters.values
          .map((f) => f.raw)
          .slice()
          .sort()
      ),
    ];
    const responseFilters = [
      ...new Set(result.filters.map((m) => m.text).sort()),
    ];
    if (JSON.stringify(queryFilters) !== JSON.stringify(responseFilters)) {
      console.info(
        `Got response with filters '${responseFilters}' while expecting results for '${queryFilters}', ignoring`
      );
      return;
    }

    for (const filter of result.filters) {
      const storedIndex = this.filters.values.findIndex(
        (f) => f.raw === filter.text
      );
      this.filters.values[storedIndex] = Object.assign(
        this.filters.values[storedIndex],
        {
          applied: true,
          isValid: filter.isValid,
          hits: filter.hits,
          name: filter.name,
          matcher: filter.matcher,
          value: filter.value,
        }
      );
    }

    let updates = {};
    for (const key of [
      "colors",
      "counters",
      "grids",
      "silences",
      "upstreams",
      "receivers",
    ]) {
      updates[key] = result[key];
    }
    this.data = Object.assign(this.data, updates);

    // before storing new version check if we need to reload
    if (
      this.info.version !== "unknown" &&
      this.info.version !== result.version
    ) {
      this.info.upgradeNeeded = true;
    }
    // update extra root level keys that are stored under 'info'
    for (const key of ["totalAlerts", "version", "authentication"]) {
      if (this.info[key] !== result[key]) {
        this.info[key] = result[key];
      }
    }

    // settings exported via API
    this.settings.values = result.settings;

    this.status.setIdle();
  });

  handleFetchError = action((err) => {
    this.status.setFailure(err);

    // reset alert counter since we won't be rendering any alerts
    this.info.totalAlerts = 0;

    // all unapplied filters should be marked applied to reset progress indicator
    this.filters.applyAllFilters();

    return { error: err };
  });
}

export {
  AlertStore,
  AlertStoreStatuses,
  FormatBackendURI,
  FormatAPIFilterQuery,
  FormatAlertsQ,
  DecodeLocationSearch,
  UpdateLocationSearch,
  NewUnappliedFilter,
};
