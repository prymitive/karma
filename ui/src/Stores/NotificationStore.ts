import { action, computed, observable, makeObservable } from "mobx";

export interface NotificationT {
  id: string;
  type: "warning" | "error";
  title: string;
  message: string;
  timestamp: Date;
  dismissed: boolean;
  autoDismissed: boolean;
  source: "alertmanager" | "system";
  sourceId?: string; // For alertmanager notifications, this would be the upstream name
  occurrenceCount: number; // Track how many times this notification has occurred
}

class NotificationStore {
  notifications: NotificationT[] = [];

  constructor() {
    makeObservable(this, {
      notifications: observable,
      addNotification: action,
      dismissNotification: action,
      dismissAllBySource: action,
      clearDismissed: action,
      activeNotifications: computed,
      dismissedNotifications: computed,
      warningCount: computed,
      errorCount: computed,
      totalActiveCount: computed,
    });
  }

  addNotification = (
    notification: Omit<
      NotificationT,
      "id" | "timestamp" | "dismissed" | "autoDismissed" | "occurrenceCount"
    >,
  ) => {
    // Check if we already have an active notification for this source
    const existingActiveIndex = this.notifications.findIndex(
      (n) =>
        n.sourceId === notification.sourceId &&
        n.source === notification.source &&
        !n.dismissed,
    );

    // Check if we have a previously dismissed notification for this source
    const existingDismissedIndex = this.notifications.findIndex(
      (n) =>
        n.sourceId === notification.sourceId &&
        n.source === notification.source &&
        n.dismissed,
    );

    if (existingActiveIndex >= 0) {
      // Update existing active notification
      const newNotification: NotificationT = {
        ...notification,
        id: this.notifications[existingActiveIndex].id,
        timestamp: new Date(),
        dismissed: false,
        autoDismissed: false,
        occurrenceCount:
          this.notifications[existingActiveIndex].occurrenceCount + 1,
      };
      this.notifications[existingActiveIndex] = newNotification;
    } else if (existingDismissedIndex >= 0) {
      // Instead of creating a new notification, increment the count on the dismissed one
      const dismissedNotification = this.notifications[existingDismissedIndex];
      dismissedNotification.occurrenceCount += 1;
      dismissedNotification.timestamp = new Date(); // Update timestamp to reflect latest occurrence
      // Note: We keep it dismissed but show the updated count in history
    } else {
      // Add new notification
      const newNotification: NotificationT = {
        ...notification,
        id: `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
        timestamp: new Date(),
        dismissed: false,
        autoDismissed: false,
        occurrenceCount: 1,
      };
      this.notifications.push(newNotification);
    }

    // Keep only last 50 notifications to prevent memory issues
    if (this.notifications.length > 50) {
      this.notifications = this.notifications.slice(-50);
    }
  };

  dismissNotification = (id: string, isAutoDismiss = false) => {
    const notification = this.notifications.find((n) => n.id === id);
    if (notification) {
      notification.dismissed = true;
      notification.autoDismissed = isAutoDismiss;
    }
  };

  dismissAllBySource = (
    source: string,
    sourceId?: string,
    isAutoDismiss = false,
  ) => {
    this.notifications.forEach((notification) => {
      if (
        notification.source === source &&
        (!sourceId || notification.sourceId === sourceId) &&
        !notification.dismissed
      ) {
        notification.dismissed = true;
        notification.autoDismissed = isAutoDismiss;
      }
    });
  };

  clearDismissed = () => {
    this.notifications = this.notifications.filter((n) => !n.dismissed);
  };

  get activeNotifications(): NotificationT[] {
    return this.notifications.filter((n) => !n.dismissed);
  }

  get dismissedNotifications(): NotificationT[] {
    return this.notifications.filter((n) => n.dismissed).slice(-20); // Keep last 20 dismissed
  }

  get warningCount(): number {
    return this.activeNotifications.filter((n) => n.type === "warning").length;
  }

  get errorCount(): number {
    return this.activeNotifications.filter((n) => n.type === "error").length;
  }

  get totalActiveCount(): number {
    return this.activeNotifications.length;
  }
}

export { NotificationStore };
export const notificationStore = new NotificationStore();
