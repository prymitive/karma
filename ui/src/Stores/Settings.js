import { action } from "mobx";
import { localStored } from "mobx-stored";

class SavedFilters {
  config = localStored(
    "savedFilters",
    {
      filters: [],
      present: false
    },
    {
      delay: 100
    }
  );

  save = action(newFilters => {
    this.config.filters = newFilters;
    this.config.present = true;
  });

  clear = action(() => {
    this.config.filters = [];
    this.config.present = false;
  });
}

class FetchConfig {
  constructor(refresh) {
    this.config = localStored(
      "fetchConfig",
      { interval: refresh },
      { delay: 100 }
    );

    this.setInterval = action(newInterval => {
      this.config.interval = newInterval;
    });
  }
}

class AlertGroupConfig {
  options = Object.freeze({
    expanded: { label: "Always expanded", value: "expanded" },
    collapsedOnMobile: {
      label: "Collapse on mobile",
      value: "collapsedOnMobile"
    },
    collapsed: { label: "Always collapsed", value: "collapsed" }
  });

  constructor(renderCount, collapseState, colorTitleBar) {
    this.config = localStored(
      "alertGroupConfig",
      {
        defaultRenderCount: renderCount,
        defaultCollapseState: collapseState,
        colorTitleBar: colorTitleBar
      },
      { delay: 100 }
    );
  }

  update = action(data => {
    for (const [key, val] of Object.entries(data)) {
      this.config[key] = val;
    }
  });
}

class SilenceFormConfig {
  config = localStored("silenceFormConfig", { author: "" }, { delay: 100 });

  saveAuthor = action(newAuthor => {
    this.config.author = newAuthor;
  });
}

class GridConfig {
  options = Object.freeze({
    default: { label: "Use defaults from karma config file", value: "default" },
    disabled: { label: "No sorting", value: "disabled" },
    startsAt: { label: "Sort by alert timestamp", value: "startsAt" },
    label: { label: "Sort by alert label", value: "label" }
  });
  constructor(groupWidth) {
    this.config = localStored(
      "alertGridConfig",
      {
        sortOrder: this.options.default.value,
        sortLabel: null,
        reverseSort: null,
        groupWidth: groupWidth
      },
      { delay: 100 }
    );
  }
}

class FilterBarConfig {
  constructor(autohide) {
    this.config = localStored(
      "filterBarConfig",
      {
        autohide: autohide
      },
      {
        delay: 100
      }
    );
  }
}

class Settings {
  constructor(defaults) {
    let defaultSettings;
    if (defaults === undefined) {
      defaultSettings = {
        Refresh: 30 * 1000 * 1000 * 1000,
        HideFiltersWhenIdle: true,
        ColorTitlebar: false,
        MinimalGroupWidth: 420,
        AlertsPerGroup: 5,
        CollapseGroups: "collapsedOnMobile"
      };
    } else {
      defaultSettings = defaults;
    }

    this.savedFilters = new SavedFilters();
    this.fetchConfig = new FetchConfig(
      defaultSettings.Refresh / 1000 / 1000 / 1000
    );
    this.alertGroupConfig = new AlertGroupConfig(
      defaultSettings.AlertsPerGroup,
      defaultSettings.CollapseGroups,
      defaultSettings.ColorTitlebar
    );
    this.gridConfig = new GridConfig(defaultSettings.MinimalGroupWidth);
    this.silenceFormConfig = new SilenceFormConfig();
    this.filterBarConfig = new FilterBarConfig(
      defaultSettings.HideFiltersWhenIdle
    );
  }
}

export { Settings };
