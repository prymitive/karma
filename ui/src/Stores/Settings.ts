import { action } from "mobx";
import { localStored } from "mobx-stored";

import { UIDefaults } from "Models/UI";
import { OptionT } from "Common/Select";

interface SavedFiltersStorage {
  filters: string[];
  present: boolean;
}
class SavedFilters {
  config: SavedFiltersStorage = localStored(
    "savedFilters",
    {
      filters: [],
      present: false,
    },
    {
      delay: 100,
    }
  );

  save = action((newFilters: string[]) => {
    this.config.filters = newFilters;
    this.config.present = true;
  });

  clear = action(() => {
    this.config.filters = [];
    this.config.present = false;
  });
}

interface FetchConfigStorage {
  interval: number;
}
class FetchConfig {
  config: FetchConfigStorage;
  setInterval: (newInterval: number) => void;

  constructor(refresh: number) {
    this.config = localStored(
      "fetchConfig",
      { interval: refresh },
      { delay: 100 }
    );

    this.setInterval = action((newInterval) => {
      this.config.interval = newInterval;
    });
  }
}

export type CollapseStateT = "expanded" | "collapsedOnMobile" | "collapsed";
interface AlertGroupConfigStorage {
  defaultRenderCount: number;
  defaultCollapseState: CollapseStateT;
  colorTitleBar: boolean;
}
class AlertGroupConfig {
  options = Object.freeze({
    expanded: { label: "Always expanded", value: "expanded" },
    collapsedOnMobile: {
      label: "Collapse on mobile",
      value: "collapsedOnMobile",
    },
    collapsed: { label: "Always collapsed", value: "collapsed" },
  });

  config: AlertGroupConfigStorage;
  setDefaultRenderCount: (val: number) => void;
  setDefaultCollapseState: (val: CollapseStateT) => void;
  setColorTitleBar: (val: boolean) => void;

  constructor(
    renderCount: number,
    collapseState: CollapseStateT,
    colorTitleBar: boolean
  ) {
    this.config = localStored(
      "alertGroupConfig",
      {
        defaultRenderCount: renderCount,
        defaultCollapseState: collapseState,
        colorTitleBar: colorTitleBar,
      },
      { delay: 100 }
    );

    this.setDefaultRenderCount = action((val: number) => {
      this.config.defaultRenderCount = val;
    });
    this.setDefaultCollapseState = action((val: CollapseStateT) => {
      this.config.defaultCollapseState = val;
    });
    this.setColorTitleBar = action((val: boolean) => {
      this.config.colorTitleBar = val;
    });
  }
}

interface SilenceFormConfigStorage {
  author: string;
}
class SilenceFormConfig {
  config: SilenceFormConfigStorage;
  saveAuthor: (newAuthor: string) => void;

  constructor() {
    this.config = localStored(
      "silenceFormConfig",
      { author: "" },
      { delay: 100 }
    );

    this.saveAuthor = action((newAuthor: string) => {
      this.config.author = newAuthor;
    });
  }
}

export type SortOrderT = "default" | "disabled" | "startsAt" | "label";
interface GridConfigStorage {
  sortOrder: SortOrderT;
  sortLabel: string | null;
  reverseSort: boolean | null;
  groupWidth: number;
}
class GridConfig {
  options = Object.freeze({
    default: { label: "Use defaults from karma config file", value: "default" },
    disabled: { label: "No sorting", value: "disabled" },
    startsAt: { label: "Sort by alert timestamp", value: "startsAt" },
    label: { label: "Sort by alert label", value: "label" },
  });

  config: GridConfigStorage;
  setSortOrder: (o: SortOrderT) => void;
  setSortLabel: (l: string) => void;
  setSortReverse: (v: boolean | null) => void;
  setGroupWidth: (w: number) => void;

  constructor(groupWidth: number) {
    this.config = localStored(
      "alertGridConfig",
      {
        sortOrder: this.options.default.value,
        sortLabel: null,
        reverseSort: null,
        groupWidth: groupWidth,
      },
      { delay: 100 }
    );

    this.setSortOrder = action((o: SortOrderT) => {
      this.config.sortOrder = o;
    });
    this.setSortLabel = action((l: string) => {
      this.config.sortLabel = l;
    });
    this.setSortReverse = action((v: boolean | null) => {
      this.config.reverseSort = v;
    });
    this.setGroupWidth = action((w: number) => {
      this.config.groupWidth = w;
    });
  }
}

interface FilterBarConfigStorage {
  autohide: boolean;
}
class FilterBarConfig {
  config: FilterBarConfigStorage;
  setAutohide: (v: boolean) => void;

  constructor(autohide: boolean) {
    this.config = localStored(
      "filterBarConfig",
      {
        autohide: autohide,
      },
      {
        delay: 100,
      }
    );
    this.setAutohide = action((v: boolean) => {
      this.config.autohide = v;
    });
  }
}

export type ThemeT = "auto" | "light" | "dark";
interface ThemeConfigStorage {
  theme: ThemeT;
  animations: boolean;
}
class ThemeConfig {
  options: { [key: string]: OptionT };
  config: ThemeConfigStorage;
  setTheme: (v: ThemeT) => void;
  setAnimations: (v: boolean) => void;

  constructor(defaultTheme: ThemeT, animations: boolean) {
    this.options = Object.freeze({
      auto: {
        label: "Automatic theme, follow browser preference",
        value: "auto",
      },
      light: { label: "Light theme", value: "light" },
      dark: { label: "Dark theme", value: "dark" },
    });

    this.config = localStored(
      "themeConfig",
      {
        theme: defaultTheme,
        animations: animations,
      },
      {
        delay: 0,
      }
    );
    this.setTheme = action((v: ThemeT) => {
      this.config.theme = v;
    });
    this.setAnimations = action((v: boolean) => {
      this.config.animations = v;
    });
  }
}

interface MultiGridConfigStorage {
  gridLabel: string;
  gridSortReverse: boolean;
}
class MultiGridConfig {
  config: MultiGridConfigStorage;
  setGridLabel: (l: string) => void;
  setGridSortReverse: (v: boolean) => void;

  constructor(gridLabel: string, gridSortReverse: boolean) {
    this.config = localStored(
      "multiGridConfig",
      {
        gridLabel: gridLabel,
        gridSortReverse: gridSortReverse,
      },
      {
        delay: 100,
      }
    );

    this.setGridLabel = action((l: string) => {
      this.config.gridLabel = l;
    });
    this.setGridSortReverse = action((v: boolean) => {
      this.config.gridSortReverse = v;
    });
  }
}

class Settings {
  savedFilters: SavedFilters;
  fetchConfig: FetchConfig;
  alertGroupConfig: AlertGroupConfig;
  gridConfig: GridConfig;
  silenceFormConfig: SilenceFormConfig;
  filterBarConfig: FilterBarConfig;
  themeConfig: ThemeConfig;
  multiGridConfig: MultiGridConfig;

  constructor(defaults: UIDefaults | null | undefined) {
    let defaultSettings: UIDefaults;
    if (defaults === undefined || defaults === null) {
      defaultSettings = {
        Refresh: 30 * 1000 * 1000 * 1000,
        HideFiltersWhenIdle: true,
        ColorTitlebar: false,
        Theme: "auto",
        Animations: true,
        MinimalGroupWidth: 420,
        AlertsPerGroup: 5,
        CollapseGroups: "collapsedOnMobile",
        MultiGridLabel: "",
        MultiGridSortReverse: false,
      };
    } else {
      defaultSettings = defaults;
    }

    this.savedFilters = new SavedFilters();
    this.fetchConfig = new FetchConfig(
      defaultSettings.Refresh / 1000 / 1000 / 1000
    );
    this.alertGroupConfig = new AlertGroupConfig(
      defaultSettings.AlertsPerGroup,
      defaultSettings.CollapseGroups,
      defaultSettings.ColorTitlebar
    );
    this.gridConfig = new GridConfig(defaultSettings.MinimalGroupWidth);
    this.silenceFormConfig = new SilenceFormConfig();
    this.filterBarConfig = new FilterBarConfig(
      defaultSettings.HideFiltersWhenIdle
    );
    this.themeConfig = new ThemeConfig(
      defaultSettings.Theme,
      defaultSettings.Animations
    );
    this.multiGridConfig = new MultiGridConfig(
      defaultSettings.MultiGridLabel,
      defaultSettings.MultiGridSortReverse
    );
  }
}

export { Settings };
