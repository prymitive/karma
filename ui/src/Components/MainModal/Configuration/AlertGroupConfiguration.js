import React, { useState } from "react";
import PropTypes from "prop-types";

import InputRange from "react-input-range";

import { Settings } from "Stores/Settings";

const AlertGroupConfiguration = ({ settingsStore }) => {
  const [defaultRenderCount, setDefaultRenderCount] = useState(
    settingsStore.alertGroupConfig.config.defaultRenderCount
  );

  const onChangeComplete = (value) => {
    settingsStore.alertGroupConfig.update({ defaultRenderCount: value });
  };

  return (
    <div className="form-group mb-0 text-center">
      <InputRange
        minValue={1}
        maxValue={10}
        step={1}
        value={defaultRenderCount}
        id="formControlRange"
        onChange={setDefaultRenderCount}
        onChangeComplete={onChangeComplete}
      />
    </div>
  );
};
AlertGroupConfiguration.propTypes = {
  settingsStore: PropTypes.instanceOf(Settings).isRequired,
};

export { AlertGroupConfiguration };
