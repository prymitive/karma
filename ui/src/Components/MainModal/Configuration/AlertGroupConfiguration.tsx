import { FC, useState } from "react";

import { Range } from "react-range";

import { Settings } from "Stores/Settings";

const AlertGroupConfiguration: FC<{
  settingsStore: Settings;
}> = ({ settingsStore }) => {
  const [defaultRenderCount, setDefaultRenderCount] = useState<number[]>([
    settingsStore.alertGroupConfig.config.defaultRenderCount,
  ]);

  const onChangeComplete = (value: number) => {
    settingsStore.alertGroupConfig.setDefaultRenderCount(value as number);
  };

  return (
    <div className="p-3 text-center">
      <Range
        step={1}
        min={1}
        max={10}
        values={defaultRenderCount}
        onChange={(values) => setDefaultRenderCount(values)}
        onFinalChange={(values) => onChangeComplete(values[0])}
        renderTrack={({ props, children }) => (
          <div className="input-range-track" {...props}>
            {children}
          </div>
        )}
        renderThumb={({ props }) => (
          <div className="input-range-thumb" {...props}>
            {defaultRenderCount}
          </div>
        )}
      />
    </div>
  );
};

export { AlertGroupConfiguration };
