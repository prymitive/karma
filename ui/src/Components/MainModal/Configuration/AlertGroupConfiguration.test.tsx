import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { Settings } from "Stores/Settings";
import { AlertGroupConfiguration } from "./AlertGroupConfiguration";

let settingsStore: Settings;

beforeEach(() => {
  settingsStore = new Settings(null);
});

const FakeConfiguration = () => {
  return mount(<AlertGroupConfiguration settingsStore={settingsStore} />);
};

describe("<AlertGroupConfiguration />", () => {
  it("matches snapshot with default values", () => {
    const tree = FakeConfiguration();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("settings are updated on completed change", () => {
    const tree = FakeConfiguration();
    expect(settingsStore.alertGroupConfig.config.defaultRenderCount).toBe(5);

    const slider = tree.find(`div.input-range-thumb`).first();
    slider.simulate("click");

    slider.simulate("keyDown", { key: "ArrowLeft", keyCode: 37 });
    slider.simulate("keyUp", { key: "ArrowLeft", keyCode: 37 });

    expect(settingsStore.alertGroupConfig.config.defaultRenderCount).toBe(4);

    slider.simulate("keyDown", { key: "ArrowRight", keyCode: 39 });
    slider.simulate("keyUp", { key: "ArrowRight", keyCode: 39 });
    slider.simulate("keyDown", { key: "ArrowRight", keyCode: 39 });
    slider.simulate("keyUp", { key: "ArrowRight", keyCode: 39 });
    expect(settingsStore.alertGroupConfig.config.defaultRenderCount).toBe(6);
  });

  it("custom interval value is rendered correctly", () => {
    settingsStore.alertGroupConfig.setDefaultRenderCount(4);
    const component = FakeConfiguration();
    expect(component.find("Range").props().values).toContain(4);
  });
});
