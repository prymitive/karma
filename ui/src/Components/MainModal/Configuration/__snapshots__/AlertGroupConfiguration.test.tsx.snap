// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertGroupConfiguration /> matches snapshot with default values 1`] = `
"
<div class=\\"form-group mb-0 text-center\\">
  <div aria-disabled=\\"false\\"
       class=\\"input-range\\"
  >
    <span class=\\"input-range__label input-range__label--min\\">
      <span class=\\"input-range__label-container\\">
        1
      </span>
    </span>
    <div class=\\"input-range__track input-range__track--background\\">
      <div style=\\"left: 0%; width: 44.44444444444444%;\\"
           class=\\"input-range__track input-range__track--active\\"
      >
      </div>
      <span class=\\"input-range__slider-container\\"
            style=\\"position: absolute; left: 44.44444444444444%;\\"
      >
        <span class=\\"input-range__label input-range__label--value\\">
          <span class=\\"input-range__label-container\\">
            5
          </span>
        </span>
        <div aria-valuemax=\\"10\\"
             aria-valuemin=\\"1\\"
             aria-valuenow=\\"5\\"
             class=\\"input-range__slider\\"
             draggable=\\"false\\"
             role=\\"slider\\"
             tabindex=\\"0\\"
        >
        </div>
      </span>
    </div>
    <span class=\\"input-range__label input-range__label--max\\">
      <span class=\\"input-range__label-container\\">
        10
      </span>
    </span>
  </div>
</div>
"
`;
